File: Python/002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        result = ListNode(0)
        tmp = result
        #每一位相加的和
        sum = 0
        while l1 != None or l2 != None:
            if l1 != None:
                sum += l1.val
                l1 = l1.next
            if l2 != None:
                sum += l2.val
                l2 = l2.next
            #计算当前结果，以及产生的进位
            tmp.next = ListNode(sum % 10)
            sum //= 10
            tmp = tmp.next
        #处理最后的进位
        if sum:
            tmp.next = ListNode(sum)
        return result.next

File: Go/002-add-two-numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    result := &ListNode {-1, nil} //最后的和链表
    res := result
    sum := 0 //每位上数字的和
    for l1 != nil || l2 != nil {
        //求每位的和
        if(l1 != nil) {
            sum += l1.Val
            l1 = l1.Next
        }
        if(l2 != nil) {
            sum += l2.Val
            l2 = l2.Next
        }
        //产生链表结点, 并处理进位
        res.Next = &ListNode {sum % 10, nil}
        sum /= 10
        res = res.Next
    }
    if sum != 0 {
        res.Next = &ListNode {sum, nil}
    }
    return result.Next
}